### PR TITLE
Add note discussing immutability to `update*` docs

### PR DESCRIPTION
**Pull request by phryneas**
### Checklist:

- [ ] If this PR contains changes to the library itself (not necessary for e.g. docs updates), please include a changeset (see [CONTRIBUTING.md](../CONTRIBUTING.md#changesets))
- [ ] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [ ] Make sure all of the significant new logic is covered by tests

It has recently come up on [twitter](https://twitter.com/franleplant/status/1621196243727704064) that none of the docs state that `updateQuery/updateFragment` updates should be made in an immutable fashion. 
In most cases, this should be caught by the fact that the data passed into the `update` function is frozen, but depending on `strict` mode being on or off, this can still be silently fail.

This adds a note for immutability in the most relevant places and links to the new React docs, that explain how to write immutable updates & potentially use `immer` for that purpose.

Note: I'm very happy for any suggestions on how to word this better, or style it as some kind of "hint".

### CODE REVIEW

**Review by benjamn — APPROVED** [review]
Longer term, I'd love if we could link to an article of our own, but I agree the React docs are useful here, and linking elsewhere also has the benefit of suggesting this is not some weird Apollo Client-specific caveat.

**Review by alessbell — APPROVED** [review]

File: docs/source/api/cache/InMemoryCache.mdx [hunk 1]
```diff
@@ -433,6 +433,8 @@ The update function of `updateQuery` takes a query's current cached value and re
 
 The returned value is automatically passed to [`writeQuery`](#writequery), which modifies the cached data.
 
+Please note update function has to calculate the new value in an immutable way. You can read more about immutable updates in [the React documentation](https://beta.reactjs.org/learn/updating-objects-in-state).
```
Comment (jerelmiller) [inline docs/source/api/cache/InMemoryCache.mdx:436]:
```suggestion
Please note the `update` function has to calculate the new value in an immutable way. You can read more about immutable updates in the [React documentation](https://beta.reactjs.org/learn/updating-objects-in-state).
```

File: docs/source/api/cache/InMemoryCache.mdx [hunk 2]
```diff
@@ -892,6 +894,8 @@ The update function of `updateFragment` takes a fragment's current cached value
 
 The returned value is automatically passed to [`writeFragment`](#writefragment), which modifies the cached data.
 
+Please note update function has to calculate the new value in an immutable way. You can read more about immutable updates in [the React documentation](https://beta.reactjs.org/learn/updating-objects-in-state).
```
Comment (jerelmiller) [inline docs/source/api/cache/InMemoryCache.mdx:897]:
```suggestion
Please note the `update` function has to calculate the new value in an immutable way. You can read more about immutable updates in the [React documentation](https://beta.reactjs.org/learn/updating-objects-in-state).
```

**Review by jerelmiller — COMMENTED** [review]
I had some minor adjustments that I think will help this read a bit better. Thanks for adding this!

**Review by jerelmiller — APPROVED** [review]

**Pull request merged**